changes/2020-07-15_raw-rsa-keyring-public-private-key-validation/change.md

```diff
@@ -0,0 +1,92 @@
+[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# Raw RSA Keyring public/private key validation
+
+## Affected Features
+
+This serves as a reference of all features that this change affects.
+
+| Feature                                                                                                                                                   |
+| --------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [Raw RSA Keyring](https://github.com/awslabs/aws-encryption-sdk-specification/blob/31b0534c4259aad365f048b73231545583389c67/framework/raw-rsa-keyring.md) |
+
+## Affected Specifications
+
+This serves as a reference of all specification documents that this change affects.
+
+| Specification                                                                                                                                             |
+| --------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [Raw RSA Keyring](https://github.com/awslabs/aws-encryption-sdk-specification/blob/31b0534c4259aad365f048b73231545583389c67/framework/raw-rsa-keyring.md) |
+
+## Affected Implementations
+
+This serves as a reference for all implementations that this change affects.
+
+| Language   | Repository                                                                            |
+| ---------- | ------------------------------------------------------------------------------------- |
+| C          | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)                   |
+| Javascript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |
+
+## Definitions
+
+- "match", "matching pair" describe an interoperable RSA public key, and RSA private key
+  (i.e. a ciphertext encrypted by the public key can be decrypted by the private key)
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+Add a requirement to the Raw RSA Keyring that if both the public key and the private key are defined,
+they MUST form a matching pair.
+
+Clarify language in OnEncrypt and OnDecrypt to require failure, instead of the vague "MUST NOT succeed".
+
+## Out of Scope
+
+- All keyrings aside from the Raw RSA Keyring
+- The manner in which the Raw RSA Keyring is build/constructed/instantiated
+- How to determine if a pair of keys match
+
+## Motivation
+
+A Raw RSA Keyring with mismatched keys is by construction unable to fulfill its decryption contract.
+As such, we should not allow such a keyring to be constructed.
```

Comment:
I'm not totally sure about these statements. We're saying we cannot allow a keyring to be constructed with mismatched pairs, but we're also saying that determining if they're matched (the way we check this) is out of scope.

I'm worried this will lead to implementation differences (or potentially some languages where we can't do this as part of the keyring initialization process.

Reply:
The problem is that the intention in `key name` is to uniquely identity the wrapping key.
So while it is possible to have have some kind of boundary keyring
that could decrypt with 1 private key only to encrypt with a non-matching pair,
to identify these things as the same makes things confusing.

I will say that if 2 customers are trying to interoperate
with a Raw RSA keyring that they have named the same thing
they MAY want to do _something_ like this.

But in this case, I would suggest that they create 2 keyrings.
One to decrypt and one to encrypt.
If these 2 keyrings have the same `name` values
but different RSA wrapping keys then the requirement is achieved
and the construction is clear.

Reply:
> We're saying we cannot allow a keyring to be constructed with mismatched pairs, but we're also saying that determining if they're matched (the way we check this) is out of scope.

The method for determining if keys match seems like an implementation detail to me.

> I'm worried this will lead to implementation differences (or potentially some languages where we can't do this as part of the keyring initialization process).

Are you concerned about different implementations allowing/disallowing certain keypairs? I don't think that's a concern, as long as each implementation checks correctly.